packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -245,12 +371,153 @@ export class CardanoWsClient extends WsProvider {
    * @param request the request.
    * @returns `true` is sent, otherwise `false`.
    */
-  private request(request: WSMessage) {
+  private request(request: WSMessage, handler?: WSHandler) {
     if (this.status !== 'connected') return false;
 
+    // Heartbeat messages do not expect a response, so they neither need a requestId, ...
+    if (Object.keys(request).length > 0)
+      // ... otherwise add requestId
+      request = { ...request, requestId: ++this.requestId };
+
     this.ws.send(JSON.stringify(request));
     this.heartbeat();
 
+    if (request.requestId && handler) this.handlers.set(request.requestId, handler);
+
     return true;
   }
+
+  private transactionsByAddresses({ addresses, blockRange, pagination }: TransactionsByAddressesArgs) {
```

Comment:
How does this find transactions that happened before connecting to ws server?

Reply:
The first time it is called it actually performs the same action the `DbSyncChainHistoryProvider.transactionsByAddresses` does.